vsp.go

```diff
@@ -0,0 +1,526 @@
+		return err
+	}
+
+	var changeOut *wire.TxOut
+	switch addr := addr.(type) {
+	case w.Address:
+		vers, script := addr.PaymentScript()
+		changeOut = &wire.TxOut{PkScript: script, Version: vers}
+	default:
+		return fmt.Errorf("failed to convert '%T' to wallet.Address", addr)
+	}
+
+	tx.TxOut = append(tx.TxOut[:0], wire.NewTxOut(int64(feeInfo.FeeAmount), feeScript))
+	feeRate := v.w.internal.RelayFee()
+	scriptSizes := make([]int, len(tx.TxIn))
+	for i := range scriptSizes {
+		scriptSizes[i] = txsizes.RedeemP2PKHSigScriptSize
+	}
+	est := txsizes.EstimateSerializeSize(scriptSizes, tx.TxOut, txsizes.P2PKHPkScriptSize)
+	change := input
+	change -= tx.TxOut[0].Value
+	change -= int64(txrules.FeeForSerializeSize(feeRate, est))
+	if !txrules.IsDustAmount(dcrutil.Amount(change), txsizes.P2PKHPkScriptSize, feeRate) {
+		changeOut.Value = change
+		tx.TxOut = append(tx.TxOut, changeOut)
+		// randomize position
+		if v.prng.coinflip() {
+			tx.TxOut[0], tx.TxOut[1] = tx.TxOut[1], tx.TxOut[0]
+		}
+	}
+
+	sigErrs, err := v.w.internal.SignTransaction(ctx, tx, txscript.SigHashAll, nil, nil, nil)
+	if err != nil {
+		log.Errorf("failed to sign transaction: %v", err)
+		for _, sigErr := range sigErrs {
+			log.Errorf("\t%v", sigErr)
+		}
+		return err
+	}
+
+	return nil
+}
+
+// PayFee receives an unsigned fee tx, signs it and make a pays fee request to
+// the vsp, so the ticket get registered.
+func (v *VSP) PayFee(ctx context.Context, ticketHash chainhash.Hash, feeTx *wire.MsgTx) error {
+	if feeTx == nil {
+		return fmt.Errorf("nil fee tx")
+	}
+
+	v.ticketToFeeLock.Lock()
+	feeInfo, exists := v.ticketsToFees[ticketHash]
+	v.ticketToFeeLock.Unlock()
+	if !exists {
+		return fmt.Errorf("call GetFeeAddress first")
+	}
+
+	votingKeyWIF, err := v.w.internal.DumpWIFPrivateKey(ctx, feeInfo.VotingAddress)
+	if err != nil {
+		log.Errorf("failed to retrieve privkey for %v: %v", feeInfo.VotingAddress, err)
+		return err
+	}
+
+	// Retrieve voting preferences
+	agendaChoices, _, err := v.w.internal.AgendaChoices(ctx, &ticketHash)
+	if err != nil {
+		log.Errorf("failed to retrieve agenda choices for %v: %v", ticketHash, err)
+		return err
+	}
+
+	voteChoices := make(map[string]string)
+	for _, agendaChoice := range agendaChoices {
+		voteChoices[agendaChoice.AgendaID] = agendaChoice.ChoiceID
+	}
+
+	var payfeeResponse PayFeeResponse
+	requestBody, err := json.Marshal(&PayFeeRequest{
+		Timestamp:   time.Now().Unix(),
+		TicketHash:  ticketHash.String(),
+		FeeTx:       txMarshaler(feeTx),
+		VotingKey:   votingKeyWIF,
+		VoteChoices: voteChoices,
+	})
+	if err != nil {
+		return err
+	}
+
+	err = v.Client.Post(ctx, "/api/v3/payfee", feeInfo.CommitmentAddress,
+		&payfeeResponse, json.RawMessage(requestBody))
+	if err != nil {
+		return err
+	}
+
+	// verify initial request matches server
+	if !bytes.Equal(requestBody, payfeeResponse.Request) {
+		log.Warnf("[PAYFee] server response has differing request: %#v != %#v",
```

Comment:
Remove spam logs...

Reply:
I'm not sure if this should be considered spam log. There are a number of places where requests are made to the vsp server and they return this error. Logging it with the differences would make it easier to debug and fix. The error message returned "server response contains differing request" is a user friendly message that could easily be used by the caller of the method.